File: src/ertmp/fourcc.c
/**
 * fourcc.c — FourCC codec registry and dispatch for Enhanced RTMP v1
 *
 * Maps FourCC codes (e.g. "avc1", "hvc1", "av01", "vp09", "Opus", "mp4a")
 * to internal codec IDs, frame type names, and provides a lookup API.
 *
 * Per E-RTMP v1 §4, the FourCC replaces the legacy 4-bit CodecID field
 * when IsExHeader is set in the VideoTagHeader or AudioTagHeader byte 0.
 */
#include "ertmp.h"
#include "librtmp2/types.h"
#include <string.h>

/* ── Video FourCC table ──────────────────────────────────────────────── */

typedef struct {
    const char   fourcc[5];      /* null-terminated, e.g. "hvc1" */
    lrtmp2_video_codec_t codec;  /* internal enum               */
    const char  *name;           /* human-readable              */
} video_fourcc_entry_t;

static const video_fourcc_entry_t s_video_fourccs[] = {
    { "avc1", LRTMP2_VIDEO_H264, "H.264/AVC"   },
    { "hvc1", LRTMP2_VIDEO_H265, "H.265/HEVC"  },
    { "av01", LRTMP2_VIDEO_AV1,  "AV1"         },
    { "vp09", LRTMP2_VIDEO_VP6,  "VP9"         },  /* VP9 maps here; no dedicated VP9 enum yet */
    {    "",  LRTMP2_VIDEO_H264, NULL           },  /* sentinel */
};

/* ── Audio FourCC table ──────────────────────────────────────────────── */

typedef struct {
    const char   fourcc[5];
    lrtmp2_audio_codec_t codec;
    const char  *name;
} audio_fourcc_entry_t;

static const audio_fourcc_entry_t s_audio_fourccs[] = {
    { "Opus", LRTMP2_AUDIO_OPUS,  "Opus"   },
    { "mp4a", LRTMP2_AUDIO_AAC,   "AAC"    },
    { "mp3 ", LRTMP2_AUDIO_MP3,   "MP3"    },   /* note trailing space per spec */
    { "ec-3", LRTMP2_AUDIO_G711_A,"Dolby Digital Plus" },  /* closest existing */
    {    "",  LRTMP2_AUDIO_AAC,   NULL      },   /* sentinel */
};

/* ── Public API ──────────────────────────────────────────────────────── */

int lrtmp2_fourcc_to_video_codec(const char *fourcc, lrtmp2_video_codec_t *out)
{
    if (!fourcc || !out) return LRTMP2_ERR_INTERNAL;
    for (const video_fourcc_entry_t *e = s_video_fourccs; e->name; e++) {
        if (memcmp(fourcc, e->fourcc, 4) == 0) {
            *out = e->codec;
            return LRTMP2_OK;
        }
    }
    *out = LRTMP2_VIDEO_H264; /* fallback — unknown codec */
    return LRTMP2_ERR_UNSUPPORTED;
}

int lrtmp2_fourcc_to_audio_codec(const char *fourcc, lrtmp2_audio_codec_t *out)
{
    if (!fourcc || !out) return LRTMP2_ERR_INTERNAL;
    for (const audio_fourcc_entry_t *e = s_audio_fourccs; e->name; e++) {
        if (memcmp(fourcc, e->fourcc, 4) == 0) {
            *out = e->codec;
            return LRTMP2_OK;
        }
    }
    *out = LRTMP2_AUDIO_AAC; /* fallback */
    return LRTMP2_ERR_UNSUPPORTED;
}

const char *lrtmp2_video_codec_to_fourcc(lrtmp2_video_codec_t codec)
{
    for (const video_fourcc_entry_t *e = s_video_fourccs; e->name; e++) {
        if (e->codec == codec) return e->fourcc;
    }
    return "avc1"; /* fallback */
}

const char *lrtmp2_audio_codec_to_fourcc(lrtmp2_audio_codec_t codec)
{
    for (const audio_fourcc_entry_t *e = s_audio_fourccs; e->name; e++) {
        if (e->codec == codec) return e->fourcc;
    }
    return "mp4a"; /* fallback */
}

const char *lrtmp2_fourcc_video_name(const char *fourcc)
{
    if (!fourcc) return NULL;
    for (const video_fourcc_entry_t *e = s_video_fourccs; e->name; e++) {
        if (memcmp(fourcc, e->fourcc, 4) == 0) return e->name;
    }
    return NULL;
}

const char *lrtmp2_fourcc_audio_name(const char *fourcc)
{
    if (!fourcc) return NULL;
    for (const audio_fourcc_entry_t *e = s_audio_fourccs; e->name; e++) {
        if (memcmp(fourcc, e->fourcc, 4) == 0) return e->name;
    }
    return NULL;
}

File: src/ertmp/fourcc.rs
//! FourCC codec registry and dispatch for Enhanced RTMP v1
//!
//! Mirrors `src/ertmp/fourcc.c`.

use crate::types::{VideoCodec, AudioCodec, Result, ErrorCode};

/* ── Video FourCC table ── */

struct VideoFourCcEntry {
    fourcc: [u8; 4],
    codec: VideoCodec,
    name: &'static str,
}

static VIDEO_FOURCCS: &[VideoFourCcEntry] = &[
    VideoFourCcEntry { fourcc: *b"avc1", codec: VideoCodec::H264, name: "H.264/AVC" },
    VideoFourCcEntry { fourcc: *b"hvc1", codec: VideoCodec::H265, name: "H.265/HEVC" },
    VideoFourCcEntry { fourcc: *b"av01", codec: VideoCodec::Av1, name: "AV1" },
    VideoFourCcEntry { fourcc: *b"vp09", codec: VideoCodec::Vp6, name: "VP9" },
];

/* ── Audio FourCC table ── */

struct AudioFourCcEntry {
    fourcc: [u8; 4],
    codec: AudioCodec,
    name: &'static str,
}

static AUDIO_FOURCCS: &[AudioFourCcEntry] = &[
    AudioFourCcEntry { fourcc: *b"Opus", codec: AudioCodec::Opus, name: "Opus" },
    AudioFourCcEntry { fourcc: *b"mp4a", codec: AudioCodec::Aac, name: "AAC" },
    AudioFourCcEntry { fourcc: *b"mp3 ", codec: AudioCodec::Mp3, name: "MP3" },
    AudioFourCcEntry { fourcc: *b"ec-3", codec: AudioCodec::G711A, name: "Dolby Digital Plus" },
];

/// Convert a FourCC string to a video codec.
pub fn fourcc_to_video_codec(fourcc: &[u8]) -> Result<VideoCodec> {
    for entry in VIDEO_FOURCCS {
        if fourcc[..4] == entry.fourcc {
            return Ok(entry.codec);
        }
    }
    Ok(VideoCodec::H264) // fallback
}

/// Convert a FourCC string to an audio codec.
pub fn fourcc_to_audio_codec(fourcc: &[u8]) -> Result<AudioCodec> {
    for entry in AUDIO_FOURCCS {
        if fourcc[..4] == entry.fourcc {
            return Ok(entry.codec);
        }
    }
    Ok(AudioCodec::Aac) // fallback
}

/// Convert a video codec to its FourCC string.
pub fn video_codec_to_fourcc(codec: VideoCodec) -> &'static str {
    for entry in VIDEO_FOURCCS {
        if entry.codec == codec {
            return std::str::from_utf8(&entry.fourcc).unwrap_or("avc1");
        }
    }
    "avc1"
}

/// Convert an audio codec to its FourCC string.
pub fn audio_codec_to_fourcc(codec: AudioCodec) -> &'static str {
    for entry in AUDIO_FOURCCS {
        if entry.codec == codec {
            return std::str::from_utf8(&entry.fourcc).unwrap_or("mp4a");
        }
    }
    "mp4a"
}

/// Get the human-readable name for a video FourCC.
pub fn fourcc_video_name(fourcc: &[u8]) -> Option<&'static str> {
    for entry in VIDEO_FOURCCS {
        if fourcc[..4] == entry.fourcc {
            return Some(entry.name);
        }
    }
    None
}

/// Get the human-readable name for an audio FourCC.
pub fn fourcc_audio_name(fourcc: &[u8]) -> Option<&'static str> {
    for entry in AUDIO_FOURCCS {
        if fourcc[..4] == entry.fourcc {
            return Some(entry.name);
        }
    }
    None
}
